Add basic signal handler and graceful exit

Samyak2 opened the pull request:
Fixes #1 

## Changes

 - Added a signal handler for SIGINT based on the [example](https://github.com/sevlyar/go-daemon/blob/master/examples/cmd/gd-signal-handling/signal-handling.go) in `go-daemon`
    - this ensures that the PID file is deleted
 - Gracefully exit scheduler based on a signal channel (called `stop`)
   - Update: the actual commands are not given a chance for graceful exit, the SSH connections get broken when the daemon exits
   - Update 2: graceful exit is again possible (see 504e7ce), this time with a timeout (default 2s) to ensure that it's not stuck stopping
 - Another signal channel (called `done`) to know when a scheduler exits after
   handling all cleanups
 - In summary: when you do `heiko stop`, heiko signals the schedulers (currently only 1) to exit, waits for them to cleanup and exit, exits the daemon while cleaning up the PID file.

## Bugs

See 8564bd8

~~Need to fix this before merging.~~

 - ~~If a non-exiting task (like `cat`) is running, heiko never~~
   ~~exits as the scheduler keeps waiting for the process to complete xD~~
    - ~~to test it without this bug, either remove that job from the config or comment out lines 19-21 in `internal/daemon/handlers.go`.~~
    - ~~perhaps we need to use context and timeout as described in this [SO answer](https://stackoverflow.com/a/58308642/11199009)~~

Samyak2 commented:
A little late now xD, but wouldn't this run `useradd` everytime the container is started or `exec`'d into?

MadhavJivrajani commented:
Yeah you're right, although since the script removes containers, it shouldn't be an issue as such, although yeah it feels super hacky. Thanks for pointing it out, I hadn't realised it earlier.

MadhavJivrajani commented:
Is this better?

```Dockerfile
FROM debian:stable-slim

# install dependencies
RUN apt update && apt upgrade -y
RUN apt install openssh-server vim sudo -y

# set password for root user
RUN echo 'root:yabe' | chpasswd
RUN sed -i 's/#PermitRootLogin prohibit-password/PermitRootLogin yes/g' /etc/ssh/sshd_config

RUN useradd -rm -d /home/test -s /bin/bash -u 1000 -G sudo test

RUN echo 'test:yabe' | chpasswd

USER test
WORKDIR /home/test

# start ssh service
ENTRYPOINT  echo yabe | sudo -S service ssh restart && bash
```
I'm not entirely sure about how you want to test it, but is it okay if the user is in the group of sudoers?

Samyak2 commented:
That looks much better!

The user being in sudoers shouldn't be a big issue. As long as we don't use commands with `sudo`, right?

MadhavJivrajani commented:
Yeah
@psiayn what do you think?

psiayn commented:
Oh yeah, makes sense!

Samyak2 committed:
add basic signal handler and graceful exit

 - add a signal handler for SIGINT based on the example in go-daemon
 - Gracefully exit scheduler based on a signal channel
 - Another signal channel to know when a scheduler exits after
   handling all cleanups
 - bug: if a non-exiting task (like `cat`) is running, heiko never
   exits as the scheduler keeps waiting for the process to complete
    - need to fix this asap

Samyak2 committed:
fix: long running comands wont prevent heiko stop

 - the goroutines that run the tasks are not waited for
 - this means that there is no chance given for graceful exit
 - the pid files are still cleaned up though

Samyak2 committed:
graceful exit: use timeout and stuff

 - here is how it works
   - a new goroutine is created to wait on the wg and signal on a new
     channel when done
   - in a for-select loop, drain the task channel and mark as done
      - with a timeout (default: 2 seconds)
      - the loop exits on timeout or when the new signal channel is
        closed
      - so it either exits when all tasks are done or timeout is reached
 - also show error in stop if SendCommands fails

psiayn reviewed (approved):
LGTM!

Samyak2 committed:
chore: ran go fmt on all files

 - used `gofmt -l -s -w .`
